lang/resources/org/partiql/type-domains/partiql.ion

```diff
@@ -426,4 +454,279 @@
+                (join
+                    join_type::join_type
+                    left::bexpr
+                    right::bexpr
+                    predicate::expr)
+
+                // Skips `row_count` rows, then emits all remaining rows.
+                (offset row_count::expr source::bexpr)
+
+                // Emits `row_count` rows, discards all remaining rows.
+                (limit row_count::expr source::bexpr)
+
+                // For every row of `source`, adds each specified `let_binding`.
+                (let source::bexpr bindings::(* let_binding 1))
+            )
+        )
+
+        // Nodes excluded below this line will eventually have a representation in the logical algebra, but not
+        // initially.
+
+        (with statement
+            (exclude
+                dml
+                ddl
+            )
+        )
+
+        (exclude
+            group_by
+            grouping_strategy
+            group_key
+            group_key_list
+            order_by
+            sort_spec
+            ordering_spec
+
+            let
+
+            dml_op
+            dml_op_list
+            ddl_op
+            conflict_action
+            on_conflict
+            returning_expr
+            returning_elem
+            column_component
+            returning_mapping
+            assignment
+            identifier
+        )
+    )
+)
+
+// partiql_logical_resolved is a variation of partiql_logical wherein all variable declarations have been allocated
+// unique identifiers and variable references have been resolved.   The first set of optimizations such as partial
+// push-downs of filters and projections may be applied to this domain.
+(define partiql_logical_resolved
+    (permute_domain partiql_logical
+        // Add `locals` to `plan`.
+        (exclude plan)
+        (include
+            (record plan
+                (stmt statement)
+                (version int)
+                (locals (* local_variable 0))
+            )
+
+            // Local variables currently include a name and register index.  In the future, something to indicate the
+            // static type of the variable may also be included here.  The index is included explicitly (instead of
+            // allowing it to be identified by ordinal position) simply to allow it to be easily identified by humans
+            // when examining plans with many local variables.
+            (product local_variable name::symbol register_index::int)
+        )
+
+
+        // For `var_decl`, `name` with `index`.  The name of the variable can still be determined by looking at the
+        // `local_variable` with the same index.
+        (exclude var_decl scope_qualifier)
+        (include
+            (product var_decl index::int)
+        )
+
+        (with expr
+            // At this point, there should be no undefined variables in the plan since the area all rewritten to
+            // dynamic lookup function call sites (if enabled).
```

Comment:
Can you please elaborate, enabled by what? `CompilerPipeline`? Also perhaps there is an assume knowledge here that requires expanding.

Reply:
I would keep references to implementation specific details such as `CompilerPipeline` (actually `PlannerPipeline`, introduced in #590) out of this document since it can/should be used to generate the ASTs of multiple partiql implementations.  However, based on that argument, dynamic lookup function call sites are also an implementation detail.  I'll remove that.